findMinSteps.cpp

```diff
@@ -0,0 +1,79 @@
+#include <iostream>
+#include <vector>
+#include <queue>
+using namespace std;
+
+
+void BFS(vector<vector<int>> maze, int rows, int columns,
+         int startX, int startY,
+         int exitX, int exitY,
+         int *ret)
+{
+    queue<pair<int, int>> q;
+    q.push(make_pair(-1, -1));
+    const int kMOVES = 4; //size of moves
+    int moves[kMOVES][2] = {{0,1},{1,0},{0,-1},{-1,0}};
+    int level = -1;
+
+    if(maze[startX][startY]==1)   {
+        q.push(make_pair(startX, startY));
+        maze[startX][startY] = -1;
+    }
+
+    while(!q.empty()){
+        int i = q.front().first;
+        int j = q.front().second;
+        q.pop();
+        if(i == -1 && j == -1) {
+            // trick 1: terminate if queue just contain <-1, -1>
```

Comment:
It takes time to understand the trick. I suggest redesign the maze data structure. e.g.
struct Cell {
  bool can_through;
  bool is_visited;
  int depth;
};
using Maze = vector<vector<Cell>>;

In this way magic number '-1' can be avoided, also maintaining the level becomes straightforward.